Fixes the races around devicemanager Allocate() and endpoint deletion.

jiayingz opened the pull request:
There is a race in predicateAdmitHandler Admit() that getNodeAnyWayFunc()
could get Node with non-zero deviceplugin resource allocatable for a
non-existing endpoint. That race can happen when a device plugin fails,
but is more likely when kubelet restarts as with the current registration
model, there is a time gap between kubelet restart and device plugin
re-registration. During this time window, even though devicemanager could
have removed the resource initially during GetCapacity() call, Kubelet
may overwrite the device plugin resource capacity/allocatable with the
old value when node update from the API server comes in later. This
could cause a pod to be started without proper device runtime config set.

To solve this problem, introduce endpointStopGracePeriod. When a device
plugin fails, don't immediately remove the endpoint but set stopTime in
its endpoint. During kubelet restart, create endpoints with stopTime set
for any checkpointed registered resource. The endpoint is considered to be
in stopGracePeriod if its stoptime is set. This allows us to track what
resources should be handled by devicemanager during the time gap.
When an endpoint's stopGracePeriod expires, we remove the endpoint and
its resource. This allows the resource to be exported through other channels
(e.g., by directly updating node status through API server) if there is such
use case. Currently endpointStopGracePeriod is set as 5 minutes.

Given that an endpoint is no longer immediately removed upon disconnection,
mark all its devices unhealthy so that we can signal the resource allocatable
change to the scheduler to avoid scheduling more pods to the node.
When a device plugin endpoint is in stopGracePeriod, pods requesting the
corresponding resource will fail admission handler.

Tested:
Ran GPUDevicePlugin e2e_node test 100 times and all passed now.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubernetes/issues/60176

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fixes the races around devicemanager Allocate() and endpoint deletion.
```

jiayingz committed:
Fixes the races around devicemanager Allocate() and endpoint deletion.

There is a race in predicateAdmitHandler Admit() that getNodeAnyWayFunc()
could get Node with non-zero deviceplugin resource allocatable for a
non-existing endpoint. That race can happen when a device plugin fails,
but is more likely when kubelet restarts as with the current registration
model, there is a time gap between kubelet restart and device plugin
re-registration. During this time window, even though devicemanager could
have removed the resource initially during GetCapacity() call, Kubelet
may overwrite the device plugin resource capacity/allocatable with the
old value when node update from the API server comes in later. This
could cause a pod to be started without proper device runtime config set.

To solve this problem, introduce endpointStopGracePeriod. When a device
plugin fails, don't immediately remove the endpoint but set stopTime in
its endpoint. During kubelet restart, create endpoints with stopTime set
for any checkpointed registered resource. The endpoint is considered to be
in stopGracePeriod if its stoptime is set. This allows us to track what
resources should be handled by devicemanager during the time gap.
When an endpoint's stopGracePeriod expires, we remove the endpoint and
its resource. This allows the resource to be exported through other channels
(e.g., by directly updating node status through API server) if there is such
use case. Currently endpointStopGracePeriod is set as 5 minutes.

Given that an endpoint is no longer immediately removed upon disconnection,
mark all its devices unhealthy so that we can signal the resource allocatable
change to the scheduler to avoid scheduling more pods to the node.
When a device plugin endpoint is in stopGracePeriod, pods requesting the
corresponding resource will fail admission handler.

jiayingz commented:
cc @vishh @RenaudWasTaken

RenaudWasTaken commented:
/area hw-accelerators

RenaudWasTaken commented on pkg/kubelet/cm/devicemanager/endpoint.go:
```diff
@@ -39,14 +39,23 @@ type endpoint interface {
+// because its device plugin fails. DeviceManager keeps the stopped endpoint in its
+// cache during this grace period to cover the time gap for the capacity change to
+// take effect.
+const endpointStopGracePeriod = time.Duration(5) * time.Minute
```
maybe move this in types.go this sounds like an important constant :)

RenaudWasTaken reviewed (commented):
Wouldn't it be simpler to issue a callback on endpoint stop  and add that timeout at the manager level?

That would remove this weird intermediary of dead but not removed state in the endpoint and simplify the guarding code you added to all the RPC calls.

You also wouldn't need to check for devices not in sync with the endpoint since you'd always have the stopGracePeriod.

And as far as re-registration is concerned by the stop callback, you would just need to replace the callback func by a noop func before stopping the endpoint.

dims commented:
/sig node

jiayingz commented:
On Tue, Mar 6, 2018 at 3:40 PM, Renaud Gaubert <notifications@github.com>
wrote:

> *@RenaudWasTaken* commented on this pull request.
>
> Wouldn't it be simpler to issue a callback on endpoint stop and add that
> timeout at the manager level?
>
> That would remove this weird intermediary of dead but not removed state in
> the endpoint and simplify the guarding code you added to all the RPC calls.
>
> You also wouldn't need to check for devices not in sync with the endpoint
> since you'd always have the stopGracePeriod.
>
> And as far as re-registration is concerned by the stop callback, you would
> just need to replace the callback func by a noop func before stopping the
> endpoint.
>
Are you suggesting to add another map in manager to track endpoint stop
grace period? I did consider that approach at the beginning but felt
we would need to manage lifecycle of another data structure and make sure
it is consistent with endpoint update.



> ------------------------------
>
> In pkg/kubelet/cm/devicemanager/endpoint.go
> <https://github.com/kubernetes/kubernetes/pull/60856#discussion_r172695319>
> :
>
> >  }
>
> +// endpointStopGracePeriod indicates the grace period after an endpoint is stopped
> +// because its device plugin fails. DeviceManager keeps the stopped endpoint in its
> +// cache during this grace period to cover the time gap for the capacity change to
> +// take effect.
> +const endpointStopGracePeriod = time.Duration(5) * time.Minute
>
> maybe move this in types.go this sounds like an important constant :)
>
done.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/60856#pullrequestreview-101757120>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AcIZlFZNnNYLg86f9LK1RsRVlEToSqATks5tbx5tgaJpZM4SfcOT>
> .
>

RenaudWasTaken commented:
> Are you suggesting to add another map in manager to track endpoint stop grace period? I did consider that approach at the beginning but felt we would need to manage lifecycle of another data structure and make sure it is consistent with endpoint update.

We probably need to move the devices in it's own structure anyways as you mentioned managing the lifcycle of these structures in the manager is complex and needs to be carefully checked.

What do you think of something like:
```golang
type ManagerStore interface {
     // This sets the timer when Update(rName, [], [], allTheDevices)
     Update(rName string, added, updated, deleted []pluginapi.Device)

     // This decides whether an a resource may be removed based on the time since the last Update
     GetCapacity() (capacity, allocatable v1.ResourceList, deleted []string)
}
```

jiayingz commented:
On Tue, Mar 6, 2018 at 4:32 PM, Renaud Gaubert <notifications@github.com>
wrote:

> Are you suggesting to add another map in manager to track endpoint stop
> grace period? I did consider that approach at the beginning but felt we
> would need to manage lifecycle of another data structure and make sure it
> is consistent with endpoint update.
>
> We probably need to move the devices in it's own structure anyways as you
> mentioned managing the lifcycle of these structures in the manager is
> complex and needs to be carefully checked.
>
> What do you think of something like:
>
> type ManagerStore interface {
>      // This sets the timer when Update(rName, [], [], allTheDevices)
>      Update(rName string, added, updated, deleted []pluginapi.Device)
>
>      // This decides whether an a resource may be removed based on the time since the last Update
>      GetCapacity() (capacity, allocatable v1.ResourceList, deleted []string)
> }
>
> Having a separate data store will perhaps help simplifying the data
structure management.
Still needs to actually try it to see how it looks.

Jiaying

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/60856#issuecomment-370979390>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AcIZlN1_W8ZKDp27NZpYB7FbltvVoZGNks5tbyqbgaJpZM4SfcOT>
> .
>

jiayingz commented:
/test pull-kubernetes-kubemark-e2e-gce

jiayingz commented:
/test pull-kubernetes-kubemark-e2e-gce

jiayingz commented:
/assign @vishh

vishh commented:
given that kubelet processed incoming pods serially during admission, this change can potentially block non device plugin pods from starting up right?
Instead should we consider a model where pods that do use only first class resources do not get impacted by device plugins?
Imagine a cluster admin trying to run a container on a specific node to exec and debug the host and the admin's debug pod not starting on the node because of this change.

jiayingz commented:
Per the offline discussion with @vishh, given that pod admission is run serially by a single process and there is a chance that the device plugin pod may even be queued behind a pod that requests the device plugin resource, it seems better for now to just fail the pod admission if it requests a disconnected device plugin resource. In 1.11, we should probably move Allocate grpc call outside of pod admission so that we can have certain retry grace period. I modified the PR to only include endpoint stopGracePeriod part so that devicemanager can properly fail pod admission during that time window. Also modified the device plugin e2e_node test to make sure we don't create pods too early after kubelet restart so that they won't fail admission. PTAL.

vishh commented:
/lgtm
/approve

dims commented:
/test pull-kubernetes-e2e-gce

vikaschoudhary16 commented:
/lgtm
